### PR TITLE
Parallel Compatible Cross-Mesh Interpolation

### PR DESCRIPTION
**Pull request by ReubenHill**
# Description
This implements cross-mesh interpolation for interpolating into function spaces which have point evaluation nodes. This is an attempt to integrate [this gist](https://gist.github.com/ReubenHill/15cb3e7fd16255f966a2cbae80250e24) into Firedrake and makes use of the recent merging of #2936.

We use `VertexOnlyMesh` as an intermediary for the global locations of the point evaluation nodes of the target function space:
 1. We get the point evaluation node locations using the method described in the [interpolation from external data section of the manual](https://www.firedrakeproject.org/interpolation.html#interpolation-from-external-data). **This will have the parallel domain decomposition of the target mesh**.
 2. Next we create a `VertexOnlyMesh` (A) at those locations **within the source mesh** such that we inherit the source mesh's parallel domain decomposition.
 3.  We interpolate our expression in our source function space onto a `P0DG` function space on `VertexOnlyMesh` (A), which has the effect of point evaluating at the target function space node locations.
 4. This `VertexOnlyMesh` (A) has an `input_ordering` `VertexOnlyMesh` (B) whose vertices have the ordering and parallel domain decomposition of the target function space global node locations. We interpolate from `P0DG` on (A) onto `P0DG` on (B). Under the hood, this is an SF reduce operation which moves the point evaluations from (A) to (B).
 5. We now have a `Function` on the `input_ordering` `VertexOnlyMesh` (B) which has point evaluations from our source mesh function space at the target mesh function space node locations. These are in the correct order and have the correct domain decomposition. We can therefore safely set the `dat` of a function in our target function space to the values of this function.

For this to work for the general case, we would need to create a `VertexOnlyMesh` at the global quadrature points of the target function space, which is rather more complicated than the work I've done here.

Some important notes:
 - This does not require one mesh to be a structured refinement of the other. This, for example, should allow you to solve a PDE on two different unstructured meshes, one of which is finer than the other, and directly check the difference in solutions by interpolating from one mesh to the other within Firedrake.
 - The target mesh merely needs to share _some_ of its domain with the source mesh and we should get the results we expect (see to do note below).
 - Crucially, this is entirely parallel compatible!

Other notes:
 - The `VertexOnlyMesh` required is stored in the `Interpolator`, as are the underlying `Interpolator`s
 - For interpolation between mixed spaces, I create `sub_interpolators` for each space and evaluate them as necessary when calling `interpolate`
   - Interpolation into a mixed space therefore requires the function space being interpolated from to be another mixed space. I throw a `NotImplementedError` if not.

To do:
- [x] Test with more meshes
  - [x] ~This ought to work with meshes of different geometric dimensions as long as the target mesh is fully within the source mesh. I should check this, because this would be really cool!~ You need the geometric dimensions to match or you would need to manually specify where the lower dimensional mesh was within the higher dimensional one! This is the whole point of immersed meshes and there's no reason why an immersed mesh wouldn't work as expected methinks.
  - [x] Check that meshes with different domains work as expected.
    - [x] ~Decide whether to optionally allow interpolation into a mesh with extent beyond the target mesh - I think this is possible with how I've implemented `VertexOnlyMesh`'s `input_ordering` property, so long as I switch off giving an error when not all the points are found, and would result in a `Function` with some values that are unchanged by the interpolation operation.~ Implemented!
- [x] Work out how to ensure we only try to interpolate into point evaluation function spaces
- [x] ~Test the adjoint - since this uses pure Firedrake, I think it ought to work fine. But perhaps setting the dat values will cause an issue?~ EDIT: Testing adjoint looks to be too tricky
- [x] Implement tests for exact interpolation and transpose interpolation by using meshes which are refinements of each other
- [ ] Document in the manual

## Associated Pull Requests:
None

## Fixes Issues:
None, this is a new feature.

# Checklist for author:

- [x] I have linted the codebase (`make lint` in the `firedrake` source directory).
- [x] My changes generate no new warnings.
- [x] All of my functions and classes have appropriate docstrings.
- [x] I have commented my code where its purpose may be unclear.
- [ ] I have included and updated any relevant documentation.
- [ ] Documentation builds locally (`make linkcheck; make html; make latexpdf` in `firedrake/docs` directory)
- [x] I have added tests specific to the issues fixed in this PR.
- [x] I have added tests that exercise the new functionality I have introduced
- [x] Tests pass locally (`pytest tests` in the `firedrake` source directory) (useful, but not essential if you don't have suitable hardware).
- [ ] I have performed a self-review of my own code using the below guidelines.

# Checklist for reviewer:

- [ ] Docstrings present
- [ ] New tests present
- [ ] Code correctly commented
- [ ] No bad "code smells"
- [ ] No issues in parallel
- [ ] No CI issues (excessive parallelism/memory usage/time/warnings generated)
- [x] Upstream/dependent branches and PRs are ready

### CODE REVIEW

File: tests/regression/test_interpolate_cross_mesh.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+from firedrake import *
+import numpy as np
+import pytest
+
+
+def test_unit_square():
```
Comment (connorjward) [inline tests/regression/test_interpolate_cross_mesh.py:6]: I think this would be better style to separate the different parts of this test into individual tests. You can use fixtures for the common setup operations.

**Review by connorjward — CHANGES_REQUESTED** [review]
Exciting!

File: tests/regression/test_interpolate_cross_mesh.py (hunk 1 above)
Comment (ReubenHill) [reply tests/regression/test_interpolate_cross_mesh.py:6]: I am in the process of doing something very similar to this. When using fixtures, does it redo the the setup every time you use the fixture? That's been my hesitation thus far.

Comment (connorjward) [reply tests/regression/test_interpolate_cross_mesh.py:6]: You can set a [fixture's scope](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) so it only gets created once. I guess that that is tricky when combined with parallel tests.

File: tests/regression/test_interpolate_cross_mesh.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                [sqrt(2) / 2, sqrt(2) / 2],
+                [-sqrt(3) / 2, 1 / 2],
+            ]
+        )  # points that ought to be on the unit circle
+        # only add target mesh vertices since they are common to both meshes
+        vertices_dest = allgather(m_dest.comm, m_dest.coordinates.dat.data_ro)
+        coords = np.concatenate((coords, vertices_dest))
+        # function.at often gets conflicting answers across boundaries for this
+        # mesh, so we lower the tolerance a bit for this test
+        m_dest.tolerance = 0.5
```
Comment (ReubenHill) [inline tests/regression/test_interpolate_cross_mesh.py:51]: Having to do this makes me wonder if the current default tolerance of `1.0` is too high.

Comment (ReubenHill) [reply tests/regression/test_interpolate_cross_mesh.py:51]: Sadly `.at` doesn't have my fancy voting algorithm to fall back on!

File: tests/regression/test_interpolate_cross_mesh.py (hunk 1 above)
Comment (ReubenHill) [reply tests/regression/test_interpolate_cross_mesh.py]: I've done a bit of refactoring to use fixtures, but I think I should probably do more for clarity. I think I'll just have to accept the extra setup cost (given this won't work for parallel tests).

**Comment by stephankramer** [comment]
Very exciting! Quick question out of interest: will this be adjointable? I'm guessing step 3 already is, step 5 should be trivial to adjoint - but step 4 requires reverse communication, is that also easy?

**Comment by dham** [comment]
> Very exciting! Quick question out of interest: will this be adjointable? I'm guessing step 3 already is, step 5 should be trivial to adjoint - but step 4 requires reverse communication, is that also easy?

Happily PETSc SF gives you the reverse communication so this stage is adjointed (already). You are correct that this means that the mesh to mesh interpolation is adjoinable.

**Comment by ReubenHill** [comment]
> Very exciting! Quick question out of interest: will this be adjointable? I'm guessing step 3 already is, step 5 should be trivial to adjoint - but step 4 requires reverse communication, is that also easy?

Once I have correctly implemented the transpose interpolation operation, this will be adjointable out of the box since interpolation is annotated. What use cases do you have in mind?

File: firedrake/interpolation.py [hunk 3]
```diff
@@ -124,31 +128,278 @@ class Interpolator(object):
...
+                ] = f_src_at_dest_node_coords_m_dest_decomp.dat.data_ro[indices]
+
+        else:
+            # We would have to set up everything again if we wanted to do this
+            # properly, but we can cheat and interpolate the other way! This
+            # won't work when we enforce that transpose interpolators can only
+            # be used with cofunctions.
+            if not self.transpose_interpolator:
+                V_dest = self.expr.function_space()
+                self.transpose_interpolator = CrossMeshInterpolator(
```
Comment (ReubenHill) [inline firedrake/interpolation.py:363]: This is just interpolating the other way, which is NOT the same as getting the adjoint/transpose interpolator.

File: firedrake/interpolation.py [hunk 4]
```diff
@@ -124,31 +128,278 @@ class Interpolator(object):
...
+            # attribute of the VOM
+            P0DG_vom = firedrake.FunctionSpace(
+                self.vom_dest_node_coords_in_m_src, "DG", 0
+            )
+            P0DG_vom_i_o = firedrake.FunctionSpace(
+                self.vom_dest_node_coords_in_m_src.input_ordering, "DG", 0
+            )
+            f_src_at_dest_node_coords_m_src_decomp = firedrake.Function(
+                P0DG_vom
+            ).interpolate(f_src)
```
Comment (ReubenHill) [inline firedrake/interpolation.py:321]: I should stash these interpolators

File: tests/regression/test_interpolate_cross_mesh.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        V_dest = FunctionSpace(m_dest, "CG", 3)
+        V_dest_2 = FunctionSpace(m_dest, "CG", 1)
+    else:
+        raise ValueError("Unknown mesh pair")
+    expr = reduce(expr_op, SpatialCoordinate(m_src))
+    expected = reduce(expr_op, coords.T)
+    return m_src, m_dest, coords, expr, expected, V_src, V_dest, V_dest_2
+
+
+def test_interpolate_cross_mesh(parameters):
```
Comment (ReubenHill) [inline tests/regression/test_interpolate_cross_mesh.py:85]: I need to add tests where we have exact refinements of meshes to check for zero error in the interpolation and transpose interpolation operations

**Comment by stephankramer** [comment]
That's fantastic! For use cases: any time-dependent models with dynamic mesh adaptivity where I want an adjoint, say for sensitivity analysis. For optimisation and goal-based adaptivity it gets a bit more involved as you have to manage the sequence of meshes in the outer loop being adapted which requires re-taping (e.g. https://github.com/pyroteus/pyroteus)

File: firedrake/interpolation.py [hunk 6]
```diff
@@ -113,6 +113,10 @@ class Interpolator(object):
         interpolation over.
     :kwarg freeze_expr: Set to True to prevent the expression being
         re-evaluated on each call.
+    :kwarg access: The access descriptor for combining updates to shared dofs.
+    :kwarg bcs: An optional list of boundary conditions to apply.
```
Comment (ReubenHill) [inline firedrake/interpolation.py:117]: Need to check if this is correct docstring

File: firedrake/interpolation.py [hunk 7]
```diff
@@ -124,31 +128,406 @@ class Interpolator(object):
...
+        self.m_src = m_src
+        self.m_dest = m_dest
+
+        self.sub_interpolators = []
+
+        # Create a VOM at the nodes of V_dest in m_src. We don't include halo
+        # node coordinates because interpolation doesn't usually include halos.
+        # NOTE: it is very important to set redundant=False, otherwise the
+        # input ordering VOM will only contain the points on rank 0!
+        # QUESTION: Should any of the below have annotation turned off?
```
Comment (ReubenHill) [inline firedrake/interpolation.py:256]: So, _should_ the below have annotation turned off?

File: firedrake/interpolation.py [hunk 8]
```diff
@@ -124,31 +128,406 @@ class Interpolator(object):
...
+        P0DG_vom_i_o = fs_type(
+            self.vom_dest_node_coords_in_m_src.input_ordering, "DG", 0
+        )
+        self.to_input_ordering_interpolator = Interpolator(
+            firedrake.TrialFunction(P0DG_vom), P0DG_vom_i_o
+        )
+        # The P0DG function outputted by the above interpolation has the
+        # correct parallel decomposition for the nodes of V_dest in m_dest so
+        # we can safely assign the dat values. This is all done in the actual
+        # interpolation method below.
```
Comment (ReubenHill) [inline firedrake/interpolation.py:347]: Should any of this have annotation turned off?

File: firedrake/interpolation.py [hunk 9]
```diff
@@ -113,6 +113,13 @@ class Interpolator(object):
         interpolation over.
     :kwarg freeze_expr: Set to True to prevent the expression being
         re-evaluated on each call.
+    :kwarg access: The access descriptor for combining updates to shared dofs.
```
Comment (connorjward) [inline firedrake/interpolation.py:116]: Add something like: "Possible values are `WRITE` and `INC`"?

File: firedrake/interpolation.py [hunk 10]
```diff
@@ -113,6 +113,13 @@ class Interpolator(object):
...
+    :kwarg bcs: An optional list of boundary conditions to zero-out in the
+        output function space. Interpolator rows or columns which are
+        associated with boundary condition nodes are zeroed out when this is
+        specified.
+    :kwarg allow_missing_points: For interpolation across meshes: allow
+        points in the target mesh that are not in the source mesh.
```
Comment (connorjward) [inline firedrake/interpolation.py:122]: What happens when this is the case? And if `False` I expect that a particular exception is raised.

File: firedrake/interpolation.py [hunk 11]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
+            optional value to assign to DOFs in the target mesh that are
+            outside the source mesh. If this is not set and an ``output``
+            :class:`.Function` is specified, then such DOF values are left
+            unchanged. If this is not set and no ``output`` :class:`.Function`
+            is specified, then the default value is zero. This does not effect
+            transpose interpolation.
+
         :returns: The resulting interpolated :class:`.Function`.
         """
+        return self.interpolate(*args, **kwargs)
```
Comment (connorjward) [inline firedrake/interpolation.py:185]: Surely this would loop infinitely? Mark as an abstract method?

File: firedrake/interpolation.py [hunk 12]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
             interpolate.
         :kwarg output: Optional. A :class:`.Function` to contain the output.
         :kwarg transpose: Set to true to apply the transpose (adjoint) of the
               interpolation operator.
+        :kwarg default_missing_val: For interpolation across meshes: the
+            optional value to assign to DOFs in the target mesh that are
+            outside the source mesh. If this is not set and an ``output``
+            :class:`.Function` is specified, then such DOF values are left
+            unchanged. If this is not set and no ``output`` :class:`.Function`
+            is specified, then the default value is zero. This does not effect
```
Comment (connorjward) [inline firedrake/interpolation.py:180]:
```suggestion
            is specified, then the default value is zero. This does not affect
```

File: firedrake/interpolation.py [hunk 13]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
         """
+        return self.interpolate(*args, **kwargs)
+
+
+class CrossMeshInterpolator(Interpolator):
+    """
+    Interpolate a function from one mesh and function space to another.
+
+    Only works when the function space we interpolate into has point evaluation
+    nodes. A more general version would require finding the global coordinates
```
Comment (connorjward) [inline firedrake/interpolation.py:193]: This is effectively a TODO comment. I don't think it belongs in the class docstring. Perhaps leave a descriptive comment where you raise the `NotImplementedError`.

File: firedrake/interpolation.py (hunk 11 above)
Comment (connorjward) [reply firedrake/interpolation.py:185]: And in fact `Interpolator` should probably inherit from `abc.ABC`:

```py
class Interpolator(abc.ABC):
    ...
```

since it should never be directly instantiated.

File: firedrake/interpolation.py [hunk 14]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
+        self,
+        expr,
+        V,
+        subset=None,
+        freeze_expr=False,
+        access=op2.WRITE,
+        bcs=None,
+        allow_missing_points=False,
+    ):
+        if subset:
```
Comment (connorjward) [inline firedrake/interpolation.py:209]: It should definitely be documented that these options will not work. The only way a user would find out that `subset` etc won't work is by trying it and having it break.

File: firedrake/interpolation.py [hunk 15]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
+
+        self.arguments = extract_arguments(expr)
+        self.nargs = len(self.arguments)
+        if self.nargs and self.arguments[0] != self.expr:
+            raise NotImplementedError(
+                "Can't yet create an interpolator from an expression with arguments."
+            )
+
+        if self._allow_missing_points:
+            missing_points_behaviour = None
```
Comment (connorjward) [inline firedrake/interpolation.py:236]: I think this would be better as an `enum`:

```py
class MissingPointsBehaviour(enum.Enum):
    NONE = enum.auto()  # not a great name
    ERROR = enum.auto()
```

File: firedrake/interpolation.py [hunk 16]
```diff
@@ -124,31 +131,406 @@ class Interpolator(object):
...
+        V_dest = V
+        m_src = extract_unique_domain(expr)
+        m_dest = V_dest.ufl_domain()
+        m_src_gdim = m_src.geometric_dimension()
+        m_dest_gdim = m_dest.geometric_dimension()
+        if m_src_gdim != m_dest_gdim:
+            raise ValueError(
+                "geometric dimensions of source and destination meshes must match"
+            )
+        self.m_src = m_src
```
Comment (connorjward) [inline firedrake/interpolation.py:250]: Very picky but I don't like this variable name. `src_mesh` for example would make it much clearer what this is.

**Review by connorjward — DISMISSED** [review]

File: firedrake/interpolation.py (hunk 12 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:180]: Thanks, will change

File: firedrake/interpolation.py (hunk 11 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:185]: Yes, you are correct.

File: firedrake/interpolation.py (hunk 13 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:193]: This seems like a pretty important thing to highlight to a user though. How about
```
    """
    Interpolate a function from one mesh and function space to another.

    Notes
    -----
    
    Only works when the function space we interpolate into has point evaluation
    nodes. A more general version would require finding the global coordinates
    of all quadrature points of the target function space in the source mesh.

    For arguments, see :class:`.Interpolator`.
    """
```

Comment (ReubenHill) [reply firedrake/interpolation.py:193]: Thinking more about it, I agree with you.

File: firedrake/interpolation.py (hunk 9 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:116]: Do `MAX` and `MIN` not work?

File: firedrake/interpolation.py (hunk 10 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:122]: At the moment you get something like
```
ValueError: 117 vertices are outside the mesh and have been removed from the VertexOnlyMesh
```
Is this sufficient or do I need to try and catch it then rethrow with something more specific?

File: firedrake/interpolation.py (hunk 9 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:116]: Oh I've found this note in `interpolate`'s docstring
```
    .. note::

       If you use an access descriptor other than ``WRITE``, the
       behaviour of interpolation is changes if interpolating into a
       function space, or an existing function. If the former, then
       the newly allocated function will be initialised with
       appropriate values (e.g. for MIN access, it will be initialised
       with MAX_FLOAT). On the other hand, if you provide a function,
       then it is assumed that its values should take part in the
       reduction (hence using MIN will compute the MIN between the
       existing values and any new values).
```

File: firedrake/interpolation.py (hunk 16 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:250]: Changed

File: firedrake/interpolation.py (hunk 15 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:236]: This is me specifying the necessary argument to `VertexOnlyMesh` (either `None` or `'error'`) which is part of the public API. I don't see how an `enum` is therefore helpful?

File: firedrake/interpolation.py (hunk 10 above)
Comment (ReubenHill) [reply firedrake/interpolation.py:122]: I've decided to rethrow something more specific.

File: docs/source/interpolation.rst [hunk 17]
```diff
@@ -108,6 +108,22 @@ in the dual space to `V` to assembled 1-forms in the dual space to
...
+use cases for this: For example, two solutions to the same problem calculated
+on meshes with different resolutions or cell shapes can be interpolated onto
+one another, or onto a third mesh, and be directly compared.
+
+Volume, surface and line integrals can be calculated by interpolating onto the
+mesh or immersed manifold which defines the volume, surface or line of
+interest in the domain, then assembling an appropriate form over the target
+mesh function space.
+
+
```
Comment (ReubenHill) [inline docs/source/interpolation.rst:126]: @dham What would be a good example to demonstrate here?

Comment (ReubenHill) [reply docs/source/interpolation.rst:126]: The line integral from the tests perhaps?

File: firedrake/interpolation.py (hunk 10 above)
Comment (connorjward) [reply firedrake/interpolation.py]: My point was that the docstring was unclear. You should just include something like "If `False` then an exception will be raised if points are located outside the source mesh". A more specific exception sounds good though.

Comment (ReubenHill) [reply firedrake/interpolation.py]: See the updated docstring

File: firedrake/interpolation.py (hunk 15 above)
Comment (connorjward) [reply firedrake/interpolation.py:289]: I think generally if you have a finite number of options for something an enum is recommended. You can actually get away with this without changing the public API:

```py
class MissingPointsBehaviour(enum.Enum):
    NONE = None
    ERROR = "error"
```

then things like `missing_points_behaviour == MissingPointsBehaviour.ERROR` will work even if the user provides the string "error".

Comment (connorjward) [reply firedrake/interpolation.py:289]: Yes. The idea is to get rid of explicit strings in the internal code. You can also do some [straightforward validation of user input](https://stackoverflow.com/questions/43634618/how-do-i-test-if-int-value-exists-in-python-enum-without-using-try-catch).

I also think that `None`/`NONE` is not a very descriptive name. Should it be `"ignore"`/`IGNORE`?

Comment (ReubenHill) [reply firedrake/interpolation.py:289]: > I also think that `None`/`NONE` is not a very descriptive name. Should it be `"ignore"`/`IGNORE`?

Well it implies "No missing points behaviour", but if you don't think it's clear then that's fair enough. It's documented at least! But again I don't want to be changing the public API (and all the tests and manual pages) at this point.

How do enums interact with a public API in general? Is the user supposed to create an enum and supply that to the constructor? Does the displayed default value change to an instance of the enum? There are very few of these in the firedrake code base, none of which seem to be public facing, hence all my questions!

Comment (ReubenHill) [reply firedrake/interpolation.py]: Okay I've added the enum in dc4f2812245b2cb745d46b3a19c96b539cfc46ed

File: firedrake/mesh.py [hunk 18]
```diff
@@ -2841,12 +2847,17 @@ def VertexOnlyMesh(mesh, vertexcoords, missing_points_behaviour='error',
         mesh, vertexcoords, tolerance=tolerance, redundant=redundant, exclude_halos=False
     )
 
-    if missing_points_behaviour:
+    if missing_points_behaviour is None:
+        missing_points_behaviour = MissingPointsBehaviour.IGNORE
+    else:
+        missing_points_behaviour = MissingPointsBehaviour(missing_points_behaviour)
```
Comment (dham) [inline firedrake/mesh.py:2853]:
```suggestion
    missing_points_behaviour = MissingPointsBehaviour(missing_points_behaviour)
```

**Review by dham — DISMISSED** [review]

**Review by JDBetteridge — APPROVED** [review]
Previously approved

File: firedrake/interpolation.py (hunk 8 above)
Comment (JDBetteridge) [reply firedrake/interpolation.py]: https://github.com/firedrakeproject/firedrake/pull/3087
